File: parsers/ocbc.py
from __future__ import annotations

import csv
import io
from datetime import datetime

import sys
import os
sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))
from categorize import Transaction


def parse(filepath: str) -> list[Transaction]:
    if filepath.lower().endswith(".pdf"):
        raise ValueError(
            f"OCBC expects a CSV export, but got a PDF: {filepath}\n"
            "Download the CSV from OCBC internet banking (Accounts > Transaction History > Export)."
        )

    with open(filepath, encoding="utf-8", errors="replace") as f:
        raw = f.read()

    try:
        start = raw.index("Transaction date")
    except ValueError:
        raise ValueError(
            f"Could not find 'Transaction date' header in {filepath}. "
            "Make sure this is an OCBC CSV export, not a PDF or other format."
        )

    reader = csv.DictReader(io.StringIO(raw[start:]))

    transactions: list[Transaction] = []
    for row in reader:
        withdrawal_raw = (row.get("Withdrawals(MYR )") or "").strip().strip('"')
        if not withdrawal_raw:
            continue

        amount = float(withdrawal_raw.replace(",", ""))
        if amount <= 0:
            continue

        date_str = (row.get("Transaction date") or "").strip()
        if not date_str:
            continue
        date = datetime.strptime(date_str, "%d/%m/%Y").strftime("%Y-%m-%d")

        description = (row.get("Description") or "").strip().replace("\\n", " ").replace("\n", " ").replace("\r", "")

        transactions.append(Transaction(
            date=date,
            description=description,
            amount=amount,
            currency="MYR",
            payment_type="Bank transfer",
            source_account="OCBC 360",
        ))

    return transactions

File: webapp/src/parsers/ocbc.js
import Papa from 'papaparse';
import { makeTransaction } from '../categorize.js';

/**
 * @param {File} file
 * @returns {Promise<import('../categorize.js').Transaction[]>}
 */
export async function parse(file) {
  if (file.name.toLowerCase().endsWith('.pdf')) {
    throw new Error(
      `OCBC expects a CSV export, but got a PDF: ${file.name}\n` +
      'Download the CSV from OCBC internet banking (Accounts > Transaction History > Export).'
    );
  }

  const raw = await file.text();

  const start = raw.indexOf('Transaction date');
  if (start === -1) {
    throw new Error(
      `Could not find 'Transaction date' header in ${file.name}. ` +
      'Make sure this is an OCBC CSV export, not a PDF or other format.'
    );
  }

  const { data } = Papa.parse(raw.slice(start), { header: true, skipEmptyLines: true });

  const transactions = [];
  for (const row of data) {
    const withdrawalRaw = (row['Withdrawals(MYR )'] || '').trim().replace(/^"|"$/g, '');
    if (!withdrawalRaw) continue;

    const amount = parseFloat(withdrawalRaw.replace(/,/g, ''));
    if (amount <= 0) continue;

    const dateStr = (row['Transaction date'] || '').trim();
    if (!dateStr) continue;
    const [d, m, y] = dateStr.split('/');
    const date = `${y}-${m}-${d}`;

    const description = (row['Description'] || '').trim().replace(/\\n/g, ' ').replace(/\n/g, ' ').replace(/\r/g, '');

    transactions.push(makeTransaction({
      date,
      description,
      amount,
      currency: 'MYR',
      payment_type: 'Bank transfer',
      source_account: 'OCBC 360',
    }));
  }

  return transactions;
}
